libtenzir/src/pipeline.cpp

```diff
@@ -177,24 +177,38 @@ auto pipeline::optimize_into_filter() const -> std::pair<expression, pipeline> {
 
 auto pipeline::optimize_into_filter(const expression& filter) const
   -> std::pair<expression, pipeline> {
-  auto opt = optimize(filter, event_order::ordered);
+  auto opt = optimize(filter, event_order::ordered,
+                      select_optimization::no_select_optimization());
   auto* pipe = dynamic_cast<pipeline*>(opt.replacement.get());
   // We know that `pipeline::optimize` yields a pipeline and a filter.
   TENZIR_ASSERT(pipe);
   TENZIR_ASSERT(opt.filter);
   return {std::move(*opt.filter), std::move(*pipe)};
 }
 
-auto pipeline::optimize(expression const& filter, event_order order) const
+auto pipeline::optimize(expression const& filter, event_order order,
+                        select_optimization const& selection) const
   -> optimize_result {
   auto current_filter = filter;
   auto current_order = order;
+  auto current_selection = selection;
   // Collect the optimized pipeline in reversed order.
   auto result = std::vector<operator_ptr>{};
   for (auto it = operators_.rbegin(); it != operators_.rend(); ++it) {
     TENZIR_ASSERT(*it);
     auto const& op = **it;
-    auto opt = op.optimize(current_filter, current_order);
+    auto opt = op.optimize(current_filter, current_order, current_selection);
+    if (opt.selection) {
+      current_selection = *opt.selection;
+    } else if (!current_selection.fields_of_interest.empty()) {
+      auto pipe = tql::parse_internal(fmt::format(
+        "select {}", fmt::join(current_selection.fields_of_interest, ", ")));
+      TENZIR_ASSERT(pipe);
+      auto ops = std::move(*pipe).unwrap();
+      TENZIR_ASSERT(ops.size() == 1);
+      result.push_back(std::move(ops[0]));
+      current_selection = select_optimization::no_select_optimization();
+    }
```

Comment:
Did you consider an alternative implementation where there is no replacement operator, i.e., `select` never gets elided and also never added back in, because generally `select` is a cheap operator? We could still push up the information so that it could become a no-op in optimized cases.

Not sure if this is worth it, I think you can judge the impact of that best.

Reply:
I see, so the idea would be to keep the `select` in place in addition to pushing up a copy of the `select`. I can see the benefit for code accuracy since rejecting optimization means the select stays in place, but I can see this proving difficult when combining `select` + `select` or `select` + `filter`. I don't think it is worth the effort especially since the `do_not_optimize` does this simply enough.